### ISSUE TITLE
About implementing conv2d with fft.

### ISSUE THREAD

**Issue by FrontMage** [RoyiAvital/StackExchangeCodes#4]
Hi, I read your answer on stack exchange and I managed to implement fast conv2d with fft using rust.

Thank you for the amazing answer.

However, I ran into some problems when I tried to implement full conv2d. I can't find the correct way to pad the image.

When conv2d is on valid mode, the image needs no padding, because the result is the same size as the image.

When conv2d is on full mode, the result is `(image_width + kernel_width -1) * (image_height + kernel_height -1)`.

Then how do I pad the image? Can you help me with this? Thanks!

**Comment by RoyiAvital** [comment]
I'd be happy if you used the answer as reference that you'd link to it in the code. 
Also please +1 :-).

Regarding your question, FFT based convolution always applies Circular Convolution. There is no way around it.
Hence if you want `(image_width + kernel_width -1) * (image_height + kernel_height -1)` output of linear convolution, pad the image to size `(image_width + kernel_width - 1 + kernel_width) * (image_height + kernel_height - 1 + kernel_height)` and crop the middle part to get the valid linear convolution of the size you're after.

**Comment by RoyiAvital** [comment]
Yep. I think I have a MATLAB code for that.
If you want, ask such question in DSP and I will answer it with MATLAB code so you'll have a reference.

I'd appreciate if you credit me in the package and the code.

**Comment by FrontMage** [comment]
Sorry for the late reply, I was working on the StackExchange reputation.

For some reason it requires 500 rep to just make a custom question tag...

Here is the question https://meta.stackexchange.com/questions/363843/to-calculate-full-convolution-with-fast-fourier-how-to-pad-the-image.

I will certainly credit you in the package description and in the code comments, great work!

**Comment by RoyiAvital** [comment]
I posted code as an answer and +1. 

Enjoy...

**Comment by FrontMage** [comment]
Thank you again! 

Although the implementation was borrowed from ZRHan's answer, I couldn't have done it without your help from the start.

I've managed to implement `conv2` and `xcorr2` using fft, and optimized them with simd, they are almost as fast as Matlab's functions.

The code is here https://github.com/FrontMage/matlab_rust.

I quoted your answer in the code comment and the `README.md` hope you don't mind.
